chunky/src/java/se/llbit/chunky/world/JavaChunk.java

```diff
@@ -0,0 +1,377 @@
+    request.add(BLOCK_ENTITIES_POST_21W43A);
+    Map<String, Tag> dataMap = getChunkTags(request);
+    // TODO: improve error handling here.
+    if (dataMap == null) {
+      throw new ChunkLoadingException(String.format("Got null data for chunk %s", this.position));
+    }
+    Tag data = tagFromMap(dataMap);
+
+    int dataVersion = data.get(DATAVERSION).intValue();
+
+    IntIntImmutablePair chunkBounds = inclusiveChunkBounds(data);
+
+    if(reuseChunkData.get() == null || reuseChunkData.get() instanceof EmptyChunkData) {
+      reuseChunkData.set(dimension.createChunkData(reuseChunkData.get(), chunkBounds.leftInt(), chunkBounds.rightInt()));
+    } else {
+      reuseChunkData.get().clear();
+    }
+    ChunkData chunkData = reuseChunkData.get(); //unwrap mutable, for ease of use
+
+    version = chunkVersion(data);
+    Tag sections = getTagFromNames(data, LEVEL_SECTIONS, SECTIONS_POST_21W39A);
+    Tag entitiesTag = data.get(LEVEL_ENTITIES);
+    Tag tileEntitiesTag = getTagFromNames(data, LEVEL_TILEENTITIES, BLOCK_ENTITIES_POST_21W43A);
+
+    BiomeDataFactory.loadBiomeData(chunkData, data, biomePalette, minY, maxY);
+    if (sections.isList()) {
+      loadBlockData(data, chunkData, palette, minY, maxY);
+
+      if (entitiesTag.isList()) {
+        for (SpecificTag tag : (ListTag) entitiesTag) {
+          if (tag.isCompoundTag())
+            chunkData.addEntity((CompoundTag) tag);
+        }
+      }
+
+      if (tileEntitiesTag.isList()) {
+        for (SpecificTag tag : (ListTag) tileEntitiesTag) {
+          if (tag.isCompoundTag())
+            chunkData.addTileEntity((CompoundTag) tag);
+        }
+      }
+    }
+
+    // post 20w45A entities
+    if (dataVersion >= DATAVERSION_20W45A) {
+      Set<String> entitiesRequest = new HashSet<>();
+      entitiesRequest.add(ENTITIES_POST_20W45A);
+
+      Map<String, Tag> entitiesMap = getEntityTags(entitiesRequest);
+      if (entitiesMap != null) {
+        entitiesTag = entitiesMap.get(".Entities");
+        if (entitiesTag.isList()) {
+          for (SpecificTag tag : (ListTag) entitiesTag) {
+            if (tag.isCompoundTag())
+              chunkData.addEntity((CompoundTag) tag);
+          }
+        }
+      }
+    }
+  }
+
+
+  /**
+   * @return The min and max blockY for a given section array
+   */
+  private IntIntImmutablePair inclusiveChunkBounds(Tag chunkData) {
+    Tag sections = getTagFromNames(chunkData, LEVEL_SECTIONS, SECTIONS_POST_21W39A);
+    int minSectionY = Integer.MAX_VALUE;
+    int maxSectionY = Integer.MIN_VALUE;
+    if (sections.isList()) {
+      for (SpecificTag section : sections.asList()) {
+        byte sectionY = (byte) section.get("Y").byteValue();
+        if (sectionY < minSectionY) {
+          minSectionY = sectionY;
+        }
+        if (sectionY > maxSectionY) {
+          maxSectionY = sectionY;
+        }
+      }
+    }
+
+    return new IntIntImmutablePair(minSectionY << 4, (maxSectionY << 4) + 15);
```

Comment:
Oh, I see… Getting the range requires the same logic internally for min and max. Fine, although the `IntIntImmutablePair` feels a bit too generic, but I'm not sure if adding a `Range` class is worth it (unless that would have more useful methods).

Reply:
Yeah I think the range class would be useful honestly, consider
```java
yMinSlider.setRangeFrom(heightRange);
mapView.setYRange(heightRange);
```
Plus a few other places